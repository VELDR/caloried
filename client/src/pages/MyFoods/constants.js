export const GET_MY_FOODS = 'MyFoods/GET_MY_FOODS';
export const SET_MY_FOODS = 'MyFoods/SET_MY_FOODS';
export const EDIT_CUSTOM_FOOD = 'MyFoods/EDIT_CUSTOM_FOOD';
export const DELETE_CUSTOM_FOOD = 'MyFoods/DELETE_CUSTOM_FOOD';
