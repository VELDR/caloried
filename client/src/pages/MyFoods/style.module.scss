@import '@styles/base';

.page {
  @include flex-center-column(2rem);
  margin: 2rem 0;

  .container {
    @include flex-center-column(1rem);
    width: 100%;
    @include apply-at-md-size {
      max-width: 40rem;
    }
    .header {
      font-weight: 700;
      font-size: $font-size-large;
    }
    .myFoods {
      width: 100%;
      border: 1px solid var(--color-text-secondary);
      border-bottom: none;
      .noResults {
        @include flex-center-column;
        padding: 2rem;
        text-align: center;
        font-size: $font-size-regular;
        border-bottom: 1px solid var(--color-text-secondary);
        @include apply-at-md-size {
          font-size: $font-size-medium;
        }
        img {
          @include box(75%);
          @include apply-at-md-size {
            @include box(20rem);
          }
        }
      }
    }
  }
  .button {
    @include flex-items-center;
    svg {
      font-size: $font-size-large;
    }
  }
}
