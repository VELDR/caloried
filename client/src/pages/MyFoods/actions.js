import { DELETE_CUSTOM_FOOD, EDIT_CUSTOM_FOOD, GET_MY_FOODS, SET_MY_FOODS } from './constants';

export const getMyFoods = (token) => ({
  type: GET_MY_FOODS,
  token,
});

export const setMyFoods = (myFoods) => ({
  type: SET_MY_FOODS,
  myFoods,
});

export const editCustomFood = (customFoodId, food, token, callback) => ({
  type: EDIT_CUSTOM_FOOD,
  customFoodId,
  food,
  token,
  callback,
});

export const deleteCustomFood = (customFoodId, token) => ({
  type: DELETE_CUSTOM_FOOD,
  customFoodId,
  token,
});
