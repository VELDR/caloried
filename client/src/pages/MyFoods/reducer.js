import { produce } from 'immer';
import { SET_MY_FOODS } from './constants';

export const initialState = {
  myFoods: [],
};

export const storedKey = [];

const myFoodsReducer = (state = initialState, action) =>
  produce(state, (draft) => {
    switch (action.type) {
      case SET_MY_FOODS:
        draft.myFoods = action.myFoods;
        break;
    }
  });

export default myFoodsReducer;
