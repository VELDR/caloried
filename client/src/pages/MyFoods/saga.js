import { setLoading, showPopup } from '@containers/App/actions';
import { deleteCustomFoodApi, editCustomFoodApi, getMyFoodsApi } from '@domain/api';
import toast from 'react-hot-toast';
import { takeLatest, call, put } from 'redux-saga/effects';
import { DELETE_CUSTOM_FOOD, EDIT_CUSTOM_FOOD, GET_MY_FOODS } from './constants';
import { getMyFoods, setMyFoods } from './actions';

function* doGetMyFoods({ token }) {
  yield put(setLoading(true));
  try {
    const response = yield call(getMyFoodsApi, token);
    yield put(setMyFoods(response, token));
  } catch (error) {
    if (error.response && error.response.data) {
      toast.error(error.response.data.message);
    } else {
      yield put(showPopup());
    }
  } finally {
    yield put(setLoading(false));
  }
}

function* doEditCustomFood({ customFoodId, food, token, callback }) {
  yield put(setLoading(true));
  try {
    const response = yield call(editCustomFoodApi, customFoodId, food, token);
    toast.success(response.message);
    yield put(getMyFoods(token));
    yield call(callback);
  } catch (error) {
    if (error.response && error.response.data) {
      toast.error(error.response.data.message);
    } else {
      yield put(showPopup());
    }
  } finally {
    yield put(setLoading(false));
  }
}

function* doDeleteCustomFood({ customFoodId, token }) {
  yield put(setLoading(true));
  try {
    const response = yield call(deleteCustomFoodApi, customFoodId, token);
    yield put(getMyFoods(token));
    toast.success(response.message);
  } catch (error) {
    if (error.response && error.response.data) {
      toast.error(error.response.data.message);
    } else {
      yield put(showPopup());
    }
  } finally {
    yield put(setLoading(false));
  }
}

export default function* myFoodsSaga() {
  yield takeLatest(GET_MY_FOODS, doGetMyFoods);
  yield takeLatest(EDIT_CUSTOM_FOOD, doEditCustomFood);
  yield takeLatest(DELETE_CUSTOM_FOOD, doDeleteCustomFood);
}
