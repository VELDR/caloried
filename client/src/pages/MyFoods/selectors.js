import { createSelector } from 'reselect';
import { initialState } from '@containers/Client/reducer';

const selectMyFoodsState = (state) => state.myFoods || initialState;

export const selectMyFoods = createSelector(selectMyFoodsState, (state) => state.myFoods);
