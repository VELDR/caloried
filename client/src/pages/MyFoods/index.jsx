import PropTypes from 'prop-types';
import { connect, useDispatch } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useEffect, useState } from 'react';
import { FormattedMessage } from 'react-intl';
import { isRoleMatch } from '@utils/authUtils';
import { formatCustomFoodForm } from '@utils/formatUtils';
import noResults from '@static/images/no-result.svg';
import { Upload } from '@mui/icons-material';

import PrimaryButton from '@components/ui/PrimaryButton';
import CustomFoodForm from '@components/CustomFoodForm';
import FoodCard from '@components/FoodCard';
import { selectToken } from '@containers/Client/selectors';
import { selectMyFoods } from './selectors';
import { getMyFoods } from './actions';

import classes from './style.module.scss';

const MyFoods = ({ myFoods, token }) => {
  const dispatch = useDispatch();
  const [selectedFood, setSelectedFood] = useState(null);
  const [isModalOpen, setIsModalOpen] = useState(false);

  const handleOpenClick = () => {
    setIsModalOpen(true);
  };

  useEffect(() => {
    if (isRoleMatch(token, 'user')) {
      dispatch(getMyFoods(token));
    }
  }, [dispatch, token]);

  const handleFoodCardClick = (food) => {
    const formattedFood = formatCustomFoodForm(food);
    setSelectedFood(formattedFood);
  };
  const handleCloseModal = () => {
    setSelectedFood(null);
    setIsModalOpen(false);
  };
  return (
    <div className={classes.page} data-testid="my-foods">
      <div className={classes.container}>
        <div className={classes.header}>
          <FormattedMessage id="app_my_foods" />
        </div>
        <div className={classes.myFoods}>
          {myFoods && myFoods.length > 0 ? (
            myFoods?.map((food, index) => (
              <FoodCard key={index} food={food} onClick={() => handleFoodCardClick(food)} />
            ))
          ) : (
            <div className={classes.noResults}>
              <img src={noResults} alt="" />
              <FormattedMessage id="app_no_custom_foods" />
            </div>
          )}
        </div>
      </div>
      <div>
        <PrimaryButton isSubmit={false} className={classes.button} onClick={handleOpenClick}>
          <Upload />
          <FormattedMessage id="app_submit_food" />
        </PrimaryButton>
      </div>
      <CustomFoodForm open={isModalOpen} onClose={handleCloseModal} token={token} />
      {selectedFood && <CustomFoodForm open onClose={handleCloseModal} token={token} food={selectedFood} />}
    </div>
  );
};

MyFoods.propTypes = {
  myFoods: PropTypes.array,
  token: PropTypes.string,
};

const mapStateToProps = createStructuredSelector({
  myFoods: selectMyFoods,
  token: selectToken,
});

export default connect(mapStateToProps)(MyFoods);
