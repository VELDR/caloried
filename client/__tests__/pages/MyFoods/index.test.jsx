import MyFoods from '@pages/MyFoods';
import { render } from '@utils/testHelper';

let wrapper;

beforeEach(() => {
  wrapper = render(<MyFoods />);
});

describe('NotFound page', () => {
  it('should render the page correctly', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('my-foods')).toBeInTheDocument();
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
