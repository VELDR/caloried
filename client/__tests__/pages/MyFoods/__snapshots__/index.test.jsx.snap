// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotFound page should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="page"
        data-testid="my-foods"
      >
        <div
          class="container"
        >
          <div
            class="header"
          >
            Makananku
          </div>
          <div
            class="myFoods"
          >
            <div
              class="noResults"
            >
              <img
                alt=""
              />
              Anda belum membuat makanan kustom apa pun. Mulailah dengan menambahkan resep favorit Anda atau makanan yang dipersonalisasi!
            </div>
          </div>
        </div>
        <div>
          <button
            class="button button"
            data-testid="primary-button"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="UploadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M5 20h14v-2H5v2zm0-10h4v6h6v-6h4l-7-7-7 7z"
              />
            </svg>
            Tambah makanan
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="page"
      data-testid="my-foods"
    >
      <div
        class="container"
      >
        <div
          class="header"
        >
          Makananku
        </div>
        <div
          class="myFoods"
        >
          <div
            class="noResults"
          >
            <img
              alt=""
            />
            Anda belum membuat makanan kustom apa pun. Mulailah dengan menambahkan resep favorit Anda atau makanan yang dipersonalisasi!
          </div>
        </div>
      </div>
      <div>
        <button
          class="button button"
          data-testid="primary-button"
          type="button"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="UploadIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M5 20h14v-2H5v2zm0-10h4v6h6v-6h4l-7-7-7 7z"
            />
          </svg>
          Tambah makanan
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
